compiler/rustc_middle/src/mir/visit.rs

```diff
@@ -297,15 +299,7 @@ macro_rules! make_mir_visitor {
                     }
                 }
 
-                // for best performance, we want to use an iterator rather
-                // than a for-loop, to avoid calling `body::Body::invalidate` for
-                // each basic block.
-                #[allow(unused_macro_rules)]
-                macro_rules! basic_blocks {
-                    (mut) => (body.basic_blocks_mut().iter_enumerated_mut());
-                    () => (body.basic_blocks().iter_enumerated());
-                }
-                for (bb, data) in basic_blocks!($($mutability)?) {
+                for (bb, data) in basic_blocks_iter!(body, $($mutability)?) {
```

Comment:
Practically speaking this is the only place that needs to be customized (no one is using `visit_location` with a mutable visitor).

Maybe provide additional entry point `visit_body_...` instead of using an associated constant?